src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+      throw new UnauthorizedException('Invalid signature');
+    }
+
+    return true;
+  }
```

Comment:
PullProofAI

The implementation of the `WebhookSecretGuard` looks good overall. It validates the webhook signature using a secret key. However, there are a few improvements that can be suggested:

1. **Error Handling**: Consider adding more specific error messages to provide better context when unauthorized exceptions are thrown.
2. **Security**: Ensure that the `Env.SECRET.WEBHOOK` value is securely stored and managed.
3. **Type Checking**: Add type checking for the `signature` variable to avoid potential runtime errors.

```diff
+     if (typeof signature !== 'string') {
+         throw new UnauthorizedException('Invalid signature format');
+     }
```

Overall, the guard seems well-structured and secure.



<!-- This is an auto-generated comment by PullProofAI -->